src/utils/regex.cc

```diff
@@ -115,6 +133,66 @@ bool Regex::searchOneMatch(const std::string& s, std::vector<SMatchCapture>& cap
     return (rc > 0);
 }
 
+bool Regex::searchGlobal(const std::string& s, std::vector<SMatchCapture>& captures) const {
+    const char *subject = s.c_str();
+
+    bool prev_match_zero_length = false;
+    int pcre_options = 0;
+    int startOffset = 0;
+
+    while (startOffset <= s.length()) {
+        int ovector[OVECCOUNT];
+        if (prev_match_zero_length) {
+            pcre_options = PCRE_NOTEMPTY_ATSTART | PCRE_ANCHORED;
+        } else {
+            pcre_options = 0; // common case
+        }
+        int rc = pcre_exec(m_pc, m_pce, subject, s.length(), startOffset, pcre_options, ovector, OVECCOUNT);
+
+        if (rc > 0) {
+            size_t firstGroupForThisFullMatch = captures.size();
+            for (int i = 0; i < rc; i++) {
+                size_t start = ovector[2*i];
+                size_t end = ovector[2*i+1];
+                size_t len = end - start;
+                if (end > s.length()) {
+                    continue;
+                }
+                SMatchCapture capture(firstGroupForThisFullMatch + i, start, len);
+                captures.push_back(capture);
+
+                if (i == 0) {
+                    if (len > 0) {
+                        // normal case; next call to pcre_exec should start after the end of the last full match string
+                        startOffset = end;
+                        prev_match_zero_length = false;
+                    } else {
+                        // zero-length match; modify next match attempt to avoid infinite loop
+                        prev_match_zero_length = true;
+                    }
+                }
+            }
+        } else {
+            if (prev_match_zero_length) {
+                // The n-1 search found a zero-length match, so we did a subsequent search
+                // with the special flags. That subsequent exec did not find a match, so now advance
+                // by one character (unless CRLF, then advance by two)
+                startOffset++;
+                if (crlfIsNewline() && (startOffset < s.length()) && (s[startOffset-1] == '\r')
+                    && (s[startOffset] == '\n')) {
+                    startOffset++;
+                }
+                prev_match_zero_length = false;
+            } else {
+                // normal case; no match on most recent scan (with options=0).  We are done.
```

Comment:
Just my 2 cents...

Here you check the value of `rc` is greater than 0 or not - if not (eg. if it's definitely 0) you treat it that it's not matched.

Referring to the [PCRE API](http://pcre.sourceforge.net/pcre.txt), the value what you're looking for is -1.

I'm sure the chances are very-very-very small to get the value 0 (considering the value of the OVECTOR, which is 900), but may be here would be good to make some error handling. _"no match on most recent"_ is only when the value of `rc` is -1. If it's 0, then it means the `ovector` out of space. And there many more different error - which also have a very-very slim chance.

Just a few thoughts...

Reply:
Hi @airween  ,

Thanks for the observation; it is at least something for us to have considered.

As you note, the OVECTOR is enormous.  It seems incredibly unlikely that anyone would compose a regex pattern with 300 capture groups (keeping in mind that repeating capture groups like '(<sub-expression>)*' only count as one).

It is true that the functionality implication here is that _if_ this use case occurs, it will be treated the same as 'no match'.  Note, that this is also the case for the standard @rx operator in v3.x -- both before and after the changes made this past summer, for example.

We may well want to make some adjustments for this edge case going forward.  Whether we want to change main behaviour in this case might be debatable, but we should at least produce some kind of debug message.  However, I think it would make more sense to adjust all relevant occurrences of this behaviour at the same time (including the much more important 'rx' operator), rather than adjusting this pull request in isolation.

For those reasons, my inclination here is to leave the current pull request as is.

Thanks again.